src/main/java/org/opentripplanner/middleware/triptracker/NotifyBusOperator.java

```diff
@@ -0,0 +1,196 @@
+                Map<String, String> busNotificationRequests = travelerPosition.trackedJourney.busNotificationMessages;
+                if (busNotificationRequests.containsKey(routeId)) {
+                    BusOpNotificationMessage busOpNotificationMessage = JsonUtils.getPOJOFromJSON(
+                        busNotificationRequests.get(routeId),
+                        BusOpNotificationMessage.class
+                    );
+                    // Changed the saved message type from notify to cancel.
+                    busOpNotificationMessage.msg_type = 0;
+                    var httpStatus = doPost(JsonUtils.toJson(busOpNotificationMessage));
+                    if (httpStatus == HttpStatus.OK_200) {
+                        travelerPosition.trackedJourney.removeNotificationMessage(routeId);
+                    } else {
+                        LOG.error("Error {} while trying to cancel notification to bus operator.", httpStatus);
+                    }
+                }
+            }
+        } catch (Exception e) {
+            LOG.error("Could not cancel notification to bus operator.", e);
+        }
+    }
+
+    /**
+     * Send notification and provide response. The service only provides the HTTP status as a response.
+     */
+    public static int doPost(String body) {
+        if (IS_TEST) {
+            // TODO figure out how to mock a static method!
+            return 200;
+        }
+        var httpResponse = HttpUtils.httpRequestRawResponse(
+            URI.create(BUS_OPERATOR_NOTIFIER_API_URL),
+            1000,
+            HttpMethod.POST,
+            BUS_OPERATOR_NOTIFIER_API_HEADERS,
+            body
+        );
+        return httpResponse.status;
+    }
+
+    /**
+     * Create post body that will be sent to bus notification API.
+     */
+    public static String createPostBody(TravelerPosition travelerPosition) {
+        return JsonUtils.toJson(new BusOpNotificationMessage(travelerPosition.currentTime, travelerPosition));
+    }
+
+    /**
+     * Make sure the leg in question is a bus transit leg.
+     */
+    public static boolean isBusLeg(Leg leg) {
+        return leg != null && leg.mode.equalsIgnoreCase("BUS") && leg.transitLeg;
+    }
+
+    /**
+     * Make sure the bus route associated with this leg supports notifying the bus operator. The 'gtfsId' is expected in
+     * the format agency_id:route_id e.g. GwinnettCountyTransit:360.
+     */
+    public static boolean supportsBusOperatorNotification(String gtfsId) {
+        return QUALIFYING_BUS_NOTIFIER_ROUTES.contains(gtfsId);
+    }
+
+    /**
+     * Has the bus driver already been notified for this journey. The driver must only be notified once.
+     */
+    public static boolean hasNotPreviouslyNotifiedBusDriverForRoute(TrackedJourney trackedJourney, String routeId) {
+        for (String notifiedRouteId : trackedJourney.busNotificationMessages.keySet()) {
+            if (notifiedRouteId.equalsIgnoreCase(routeId)) {
```

Comment:
@binh-dam-ibigroup I'm not sure I understand. Are you staying that if a notification has previously been cancelled allow a second notification to be sent?

Reply:
@br648 Are both notifications and cancelations recorded in `trackedJourney.busNotificationMessages`?

Reply:
@binh-dam-ibigroup No, just the notification. The only difference would be the `msg_type` (1 = Notify, 0 = Cancel). When cancelling, the saved notification is retrived, the `msg_type` is changed to 0 and the request is sent. The saved notification is not updated.

Reply:
So there is no record of whether a cancel request was sent?

Reply:
That is correct. It is an easy fix to update the saved message with the updated cancelled `msg_type`. Shall I implement this?

Also, if I do implement and the notification has been previously cancelled, should I allow the notification to be sent again?

Reply:
@br648 Yes go ahead and update the notification, so that the cancel notification is sent once only. I can bring this up with the bus software folks to see if they support repeat requests.